docs/0002-events-architecture.md

```diff
@@ -0,0 +1,124 @@
+   - **Why we opted against this**: Hardcoding tracking logic within the Visual Editor would tightly couple the analytics tracking with the component, making it difficult to maintain and update. 
+     This approach would also violate the separation of concerns principle and increase the complexity of the Visual Editor codebase.
+
+3. **Implementing an Event Emitter System**:
+   - The chosen approach is to implement an Event Emitter system within the Visual Editor. This allows the component to emit events for user interactions such as button clicks and select changes. 
+     These events can be listened to by the Publishing Apps that imported the Visual Editor to perform actions like GA4 tracking.
+   - **Why we chose this option**: This approach decouples the GA4 tracking logic from the Visual Editor, making it easier to maintain and update. 
+     It also provides flexibility to handle different types of events without modifying the core Visual Editor component. 
+     This method adheres to the separation of concerns principle and simplifies the component logic.
+
+
+## Decision
+
+We will implement an Event Emitter system in the Visual Editor. The component will emit events for user interactions such as button clicks and dropdown selections. 
+These events will carry the bare minimum information that is needed by the Publishing Apps, such as button text and select text, 
+and the Publishing Apps can listen to these events to perform actions like GA tracking.
+
+The following steps outline the implementation:
+
+1. Create an `EventEmitter` class to manage event listeners and emit events.
+2. Integrate the `EventEmitter` into the `VisualEditor` component.
+3. Emit events from the `VisualEditor` component when user interactions occur (e.g., `clickButton`, `selectOption`).
+4. Publishing Apps can register event listeners to handle these events and push data to the GA data layer as needed.
+
+
+```mermaid
+classDiagram
+    class EventEmitter {
+        -events: Object
+        +subscribe(event: String, listener: Function): void
+        +unsubscribe(event: String, listenerToRemove: Function): void
+        -emit(event: String, data: Object): void
+    }
+```
+
+
+Draft pseudocode examples:
+
+### Option 1: passing a Hooks object with the events and callbacks
+
+```javascript
+const Hooks = {
+  onClick(el) {
+    console.log("click event on element", el)
+  },
+  onSelectOption(el) {
+    window.GovUK.analytics.track({"type": "selected option", "action": el.value()})
+  }
+}
+
+
+const editor = new GovspeakVisualEditor(
+  document.querySelector("#content"),
+  document.querySelector("#editor"),
+  document.querySelector("#govspeak"),
+  hooks: Hooks
+);
+```
+
+### Option 2: Actively subscribe to the events
+
+```javascript
+const editor = new GovspeakVisualEditor(
+  document.querySelector("#content"),
+  document.querySelector("#editor"),
+  document.querySelector("#govspeak"),
+);
+
+editor.subscribe("onClick", (el) => {
```

Comment:
Would it be sensible to unify clicks and select options to something like `menuItemActivated` so that we don't worry so much about the implementation specifics?
Although I suppose theres a question of if the GA schema is/should be different for the two

Reply:
I think there might be, but I'd say it probably lays outside of the scope of this ADR? For me we might want to name our events around Prosemirror commands rather than native JS events, as it may be we want to respond to something that isn't associated with a DOM event at some point. But I think we can discuss and develop this at a later point as we understand more about what the need is.

Reply:
Sounds good, I like the idea of basing events on prosemirror commands - helps with keyboard activation as well - but agree that deciding isn't essential to this ADR